File: frontend/src/api/http/readiness.ts
/** Readiness diagnostics — subsystem health summary, LLM probe, demo run. */

import { httpGet, httpPost, invokeWithError } from './_utils'
import { ErrorCode } from '@/utils/errors'

export interface ReadinessLibrary {
  configured: boolean
  path: string | null
  exists: boolean
  writable: boolean
  error: string | null
}

export interface ReadinessDatabase {
  ok: boolean
  error: string | null
}

export type ReadinessModelStatus = 'ready' | 'partial' | 'missing' | 'error' | (string & {})

export interface ReadinessModel {
  id: string
  name: string
  status: ReadinessModelStatus
  local_path: string | null
  size_mb: number | null
  expected_size_mb: number | null
  cache_dir: string | null
  last_error: string | null
}

export interface ReadinessLlm {
  configured: boolean
  provider: string | null
  model: string | null
  base_url: string | null
  has_api_key: boolean
}

export interface ReadinessOcr {
  chain: string[]
  error: string | null
}

export interface ReadinessSummary {
  library: ReadinessLibrary
  database: ReadinessDatabase
  models: ReadinessModel[]
  llm: ReadinessLlm
  ocr: ReadinessOcr
}

export interface ProbeLlmResult {
  ok: boolean
  latency_ms: number | null
  error: string | null
  provider: string
  model: string
}

export interface DemoRunResult {
  ok: boolean
  task_id: string | null
  file_path: string
  error: string | null
}

/** Aggregate readiness summary across library, DB, models, LLM and OCR. */
export async function readinessSummary(): Promise<ReadinessSummary> {
  return invokeWithError(
    () => httpGet<ReadinessSummary>('/api/v1/diagnostics/summary'),
    ErrorCode.ApiError,
  )
}

/** Probe the configured LLM endpoint and measure latency. */
export async function readinessProbeLlm(): Promise<ProbeLlmResult> {
  return invokeWithError(
    () => httpPost<ProbeLlmResult>('/api/v1/diagnostics/probe-llm'),
    ErrorCode.ApiError,
  )
}

/** Enqueue a small demo document through the pipeline as a smoke test. */
export async function readinessDemoRun(): Promise<DemoRunResult> {
  return invokeWithError(
    () => httpPost<DemoRunResult>('/api/v1/diagnostics/demo-run'),
    ErrorCode.ApiError,
  )
}

File: src/mbforge/models/readiness.py
"""Pydantic models for the readiness diagnostics endpoints.

These schemas expose the health of the library root, SQLite database,
downloadable models, LLM provider, and OCR fallback chain.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class LibraryReadiness(BaseModel):
    """Readiness of the configured library root directory."""

    configured: bool = False
    path: str | None = None
    exists: bool = False
    writable: bool = False
    error: str | None = None


class DatabaseReadiness(BaseModel):
    """Readiness of the library SQLite database."""

    ok: bool = False
    error: str | None = None


class ModelReadiness(BaseModel):
    """Readiness of a single downloadable model resource."""

    id: str
    name: str
    status: str
    local_path: str | None = None
    size_mb: float | None = None
    expected_size_mb: float | None = None
    cache_dir: str | None = None
    last_error: str | None = None


class LLMReadiness(BaseModel):
    """Readiness of the configured LLM provider (never leaks the API key)."""

    configured: bool = False
    provider: str | None = None
    model: str | None = None
    base_url: str | None = None
    has_api_key: bool = False


class OCRReadiness(BaseModel):
    """Readiness of the OCR fallback chain."""

    chain: list[str] = Field(default_factory=list)
    error: str | None = None


class ReadinessSummaryResponse(BaseModel):
    """Aggregated read-only readiness summary."""

    library: LibraryReadiness
    database: DatabaseReadiness
    models: list[ModelReadiness] = Field(default_factory=list)
    llm: LLMReadiness
    ocr: OCRReadiness


class LLMProbeResponse(BaseModel):
    """Result of a live LLM probe request."""

    ok: bool = False
    latency_ms: int | None = None
    error: str | None = None
    provider: str = ""
    model: str = ""


class DemoRunResponse(BaseModel):
    """Result of enqueueing a generated demo PDF through the pipeline."""

    ok: bool = False
    task_id: str | None = None
    file_path: str = ""
    doc_id: str | None = None
    error: str | None = None
